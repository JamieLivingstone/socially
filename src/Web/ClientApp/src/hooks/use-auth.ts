import axios from 'axios';
import { createContext, useContext, useEffect, useState } from 'react';
import { useCookies } from 'react-cookie';

type Account = {
  name: string;
  username: string;
  email: string;
};

export function useProvideAuth() {
  const [account, setAccount] = useState<Account | null>(null);
  const [initializing, setInitializing] = useState(true);
  const [cookies, setCookie, deleteCookie] = useCookies(['token']);

  useEffect(() => {
    (async function initialize() {
      if (cookies.token) {
        await fetchAccount(cookies.token);
      }

      setInitializing(false);
    })();
  }, []);

  async function fetchAccount(token: string) {
    try {
      axios.defaults.headers.common = { Authorization: `Bearer ${token}` };

      const { data } = await axios.get('/api/v1/accounts/current');

      setAccount(data);
    } catch (_) {
      logout();
    }
  }

  async function login(payload: { username: string; password: string }) {
    const { data } = await axios.post<{ token: string }>('/api/v1/accounts/login', payload);

    setCookie('token', data.token, {
      path: '/',
      sameSite: 'strict',
      maxAge: 604800, // 7 days
    });

    await fetchAccount(data.token);
  }

  async function register(payload: { name: string; username: string; email: string; password: string }) {
    const { data } = await axios.post<{ token: string }>('/api/v1/accounts/register', payload);

    setCookie('token', data.token, {
      path: '/',
      sameSite: 'strict',
      maxAge: 604800, // 7 days
    });

    await fetchAccount(data.token);
  }

  function logout() {
    deleteCookie('token', { path: '/' });
    setAccount(null);
  }

  return {
    account,
    initializing,
    login,
    register,
    logout,
  };
}

export const AuthContext = createContext<ReturnType<typeof useProvideAuth> | null>(null);

export function useAuth() {
  const context = useContext(AuthContext);

  if (!context) {
    throw new Error('useAuth must be within an AuthProvider');
  }

  return context;
}
