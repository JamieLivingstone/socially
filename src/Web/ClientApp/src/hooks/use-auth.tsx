import axios from 'axios';
import React, { createContext, useContext, useEffect, useState } from 'react';
import { useCookies } from 'react-cookie';
import { useNavigate } from 'react-router-dom';

import { routes } from '../constants';
import { Loading } from '../screens/loading';

type Account = {
  name: string;
  username: string;
  email: string;
};

function useProvideAuth() {
  const [account, setAccount] = useState<Account | null>(null);
  const [initializing, setInitializing] = useState(false);
  const [cookies, setCookie, deleteCookie] = useCookies(['token']);
  const navigate = useNavigate();

  useEffect(() => {
    (async function initialize() {
      if (cookies.token) {
        setInitializing(true);
        await fetchAccount(cookies.token);
        setInitializing(false);
      }
    })();
  }, []);

  async function fetchAccount(token: string) {
    try {
      axios.defaults.headers.common = {
        Authorization: `Bearer ${token}`,
      };

      const { data } = await axios.get('/api/v1/accounts/current');

      setAccount(data);
    } catch (_) {
      logout();
    }
  }

  async function login(payload: { username: string; password: string }) {
    const { data } = await axios.post<{ token: string }>('/api/v1/accounts/login', payload);

    setCookie('token', data.token, { secure: true, sameSite: 'strict' });
    await fetchAccount(data.token);
  }

  async function register(payload: { name: string; username: string; email: string; password: string }) {
    const { data } = await axios.post<{ token: string }>('/api/v1/accounts/register', payload);

    setCookie('token', data.token, { secure: true, sameSite: 'strict' });
    await fetchAccount(data.token);
  }

  function logout() {
    deleteCookie('token');
    setAccount(null);
    navigate(routes.HOME, { replace: true });
  }

  return {
    account,
    initializing,
    login,
    register,
    logout,
  };
}

const authContext = createContext<ReturnType<typeof useProvideAuth> | null>(null);

export function AuthProvider({ children }: { children: React.ReactNode }) {
  const auth = useProvideAuth();

  return <authContext.Provider value={auth}>{auth.initializing ? <Loading /> : children}</authContext.Provider>;
}

export function useAuth() {
  const context = useContext(authContext);

  if (!context) {
    throw new Error('useAuth must be within an AuthProvider');
  }

  return context;
}
