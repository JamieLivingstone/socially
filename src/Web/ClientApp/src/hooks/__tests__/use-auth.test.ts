import { act, renderHook } from '@testing-library/react-hooks';
import axios from 'axios';
import moxios from 'moxios';

import Providers from '@providers';

import { useAuth } from '../use-auth';

describe('useAuth', () => {
  beforeEach(() => {
    moxios.stubRequest('/api/v1/accounts/current', {
      status: 200,
      response: {
        name: 'John Doe',
        username: 'john',
        email: 'john@test.com',
      },
    });
  });

  afterEach(() => {
    document.cookie = 'token=';
  });

  describe('new session', () => {
    test('account is null', async () => {
      const { result } = renderHook(() => useAuth(), { wrapper: Providers });

      expect(result.current.account).toBeNull();
    });
  });

  describe('existing session', () => {
    test('fetches account', async () => {
      document.cookie = 'token=mock-token;';

      const { result, waitForNextUpdate } = renderHook(() => useAuth(), { wrapper: Providers });

      await waitForNextUpdate();

      expect(axios.defaults.headers.common.Authorization).toEqual('Bearer mock-token');
      expect(result.current.account).toMatchSnapshot();
    });
  });

  describe('login', () => {
    describe('given valid credentials', () => {
      test('sets auth token and fetches current account', async () => {
        moxios.stubRequest('/api/v1/accounts/login', {
          status: 200,
          response: { token: 'mock-token' },
        });

        const { result } = renderHook(() => useAuth(), { wrapper: Providers });

        expect(result.current.account).toBeNull();

        await act(() => result.current.login({ username: 'user', password: 'password' }));

        expect(result.current.account).not.toBeNull();
      });
    });
  });

  describe('register', () => {
    describe('given a valid payload', () => {
      test('sets auth token and fetches current account', async () => {
        moxios.stubRequest('/api/v1/accounts/register', {
          status: 201,
          response: { token: 'mock-token' },
        });

        const { result } = renderHook(() => useAuth(), { wrapper: Providers });

        expect(result.current.account).toBeNull();

        await act(() =>
          result.current.register({
            email: 'john@test.com',
            name: 'John Doe',
            password: 'password',
            username: 'john',
          }),
        );

        expect(result.current.account).not.toBeNull();
      });
    });
  });

  describe('logout', () => {
    test('removes cookie and sets account to null', async () => {
      document.cookie = 'token=mock-auth-token;';

      const { result, waitForNextUpdate } = renderHook(() => useAuth(), { wrapper: Providers });

      await waitForNextUpdate();

      expect(result.current.account).not.toBeNull();

      act(() => result.current.logout());

      expect(document.cookie).toEqual('');
      expect(result.current.account).toBeNull();
    });
  });
});
