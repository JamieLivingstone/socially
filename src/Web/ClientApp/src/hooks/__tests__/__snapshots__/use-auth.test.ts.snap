// Vitest Snapshot v1

exports[`useAuth > existing session > fetches account 1`] = `
{
  "email": "john@test.com",
  "name": "John Doe",
  "username": "john",
}
`;
